Tests/GRPCHTTP2TransportTests/ControlService.swift

```diff
@@ -0,0 +1,194 @@
+  func unary(
+    request: ServerRequest.Stream<Control.Method.Unary.Input>
+  ) async throws -> ServerResponse.Stream<Control.Method.Unary.Output> {
+    try await self.handle(request: request)
+  }
+
+  func serverStream(
+    request: ServerRequest.Stream<Control.Method.ServerStream.Input>
+  ) async throws -> ServerResponse.Stream<Control.Method.ServerStream.Output> {
+    try await self.handle(request: request)
+  }
+
+  func clientStream(
+    request: ServerRequest.Stream<Control.Method.ClientStream.Input>
+  ) async throws -> ServerResponse.Stream<Control.Method.ClientStream.Output> {
+    try await self.handle(request: request)
+  }
+
+  func bidiStream(
+    request: ServerRequest.Stream<Control.Method.BidiStream.Input>
+  ) async throws -> ServerResponse.Stream<Control.Method.BidiStream.Output> {
+    try await self.handle(request: request)
+  }
+}
+
+extension ControlService {
+  private func handle(
+    request: ServerRequest.Stream<ControlInput>
+  ) async throws -> ServerResponse.Stream<ControlOutput> {
+    var iterator = request.messages.makeAsyncIterator()
+
+    guard let message = try await iterator.next() else {
+      // Empty input stream, empty output stream.
+      return ServerResponse.Stream { _ in [:] }
+    }
+
+    // Check if the request is for a trailers-only response.
+    if message.hasStatus, message.isTrailersOnly {
+      let trailers = message.echoMetadataInTrailers ? request.metadata.echo() : [:]
+      let code = Status.Code(rawValue: message.status.code.rawValue).flatMap { RPCError.Code($0) }
+
+      if let code = code {
+        throw RPCError(code: code, message: message.status.message, metadata: trailers)
+      } else {
+        // Invalid code, the request is invalid, so throw an appropriate error.
+        throw RPCError(
+          code: .invalidArgument,
+          message: "Trailers only response must use a non-OK status code"
+        )
+      }
+    }
+
+    // Not a trailers-only response. Should the metadata be echo'd back?
+    let metadata = message.echoMetadataInHeaders ? request.metadata.echo() : [:]
+
+    // The iterator needs to be transferred into the response. This is okay: we won't touch the
+    // iterator again from the current concurrency domain.
+    let transfer = UnsafeTransfer(iterator)
+
+    return ServerResponse.Stream(metadata: metadata) { writer in
+      // Finish dealing with the first message.
+      switch try await self.processMessage(message, metadata: request.metadata, writer: writer) {
+      case .return(let metadata):
+        return metadata
+      case .continue:
+        ()
+      }
+
+      var iterator = transfer.wrappedValue
+      // Process the rest of the messages.
+      while let message = try await iterator.next() {
+        switch try await self.processMessage(message, metadata: request.metadata, writer: writer) {
+        case .return(let metadata):
+          return metadata
+        case .continue:
+          ()
+        }
+      }
```

Comment:
Nit, but we could replace this with a `repeat-while` loop to avoid some duplication

Reply:
Hmm, not sure this works because we need `message` to be non-nil inside the loop, with a repeat-while we'd have to `!`-it.

Reply:
Yeah you're right, it's also not possible to do optional binding here so it's okay, ignore this